### PR TITLE
Enhance logging messages with rich

### LINKED ISSUE #386: Nicer looking logging
It would be nice if Mephisto could have some better formatted logging. 

I came across this library, which has a drop-in logging handler for formatting:

https://github.com/willmcgugan/rich#logging-handler

image

I tried to implement this in `logger_core.py` and was able to get it to work to some extent, but had issues like certain log outputs were getting left out, and other log outputs were getting duplicated (printed both richly and plainly), so maybe someone else can take a look into this. Jack mentioned that the 1.1 release of Hydra should also make logging a bit easier.

### LINKED ISSUE #388: Better CLI output
It would be great to think of Mephisto's output from a user-first perspective. Here's a mockup of what a more friendly output stream could look like. In this case, I move away from just using logging as output as mentioned in #386 and focus on a more first-class support for CLI output.

### GIF:

rich_print

### Still image:

Screen Shot 2021-02-03 at 4 32 52 PM

---

### Mockup source:

Requires the `rich` package, `pip install rich`:

```python
from time import sleep
from rich.console import Console
from rich.prompt import Confirm

console = Console()
DELAY_OFF = 0

def print_summary(task):
    from rich.table import Table, Column
    from rich import box
    import locale
    locale.setlocale(locale.LC_ALL, 'en_US.UTF-8') 

    table = Table("property", Column("value", style="magenta bold"), "description", title="Task Launch Preview", box=box.ROUNDED, expand=True, show_lines=True)
    # table = Table("task_name", "blueprint", "architect", "crowd provider", "requester", title="Task Launch Preview")
    # table.add_row("my_custom_task", "simple_static_task", "local", "mturk_sanbox", "demo")
    table.add_row("task_name", "my_custom_task", "ID used to later query all HITs launched in this run.\nConfigure via: task.task_name.")
    table.add_row("blueprint", "simple_static_task", "The base template for your task. Can be found in the folder: mephisto/abstractions/blueprints.")
    table.add_row("architect", "local", "Where this task will be hosted/launched from. Configure via: mephisto/architect=...")
    table.add_row("crowd provider", "mturk_sandbox", "The underlying platform that supplies the crowdworkers (e.g. mturk, mturk_sandbox).")
    table.add_row("requester", "demo", "The account/budget to run the task through. Configure via: mephisto.provider.requester_name=... Or use the mephisto CLI for more info: 'mephisto requesters'")


    console.print(table)

    billing_table = Table("num HITs", "price", "estimated total", title="Estimated Costs")
    billing_table.add_row(str(task['hits']), locale.currency(0.02), f"[bold green]{locale.currency(task['hits'] * 0.02)}[/bold green]")
    console.print(billing_table)

def log(category, message):
    return f"[dim]{category}[/dim] - {message}"

def prepare_router():
    console.log(log("router/install", "npm install (hidden, show with 'build.show_npm_output: true')"))
    sleep(1 * DELAY_OFF)
    console.log(log("router/install", "npm build (hidden, show with 'build.show_npm_output: true')"))
    sleep(2 * DELAY_OFF)
    console.log(log("router/install", "completed npm build"))
    console.log(log("operator/launch", "created a task run under name: 'react-static-task-example'"))

def prepare_architect():
    console.log(log("architect/deploy", "building payload"))
    sleep(1 * DELAY_OFF)
    console.log(log("architect/deploy", "sending payload to final destination"))
    sleep(2 * DELAY_OFF)
    console.log(log("architect/deploy", "deployed successfully"))





def launch_tasks(console):
    from rich.emoji import Emoji
    preview_url = "http://localhost:3000"
    assignment_url = "http://localhost:3000/?worker_id=x&assignment_id="
    for job in range(3):
        # console.log(f":rocket: Launched Assignment {job+1}. Preview: '{preview_url}' | Assignment: '{assignment_url}{job+1}'")
        console.log(log(f":rocket: assignment/launch", f"assignment ID: {job+1}\nPreview: '{preview_url}'\nAssignment: '{assignment_url}{job+1}'"))
        sleep(0.1)
    sleep(2 * DELAY_OFF)


def start():
    task = { "hits": 500}
    print_summary(task)
    confirm = Confirm.ask(f"[bold]Are you sure you would like to proceed launching {task['hits']} HITs?[/bold]")
    if not confirm:
        return

    with console.status("[bold green]Setting up router...") as status:
        prepare_router()
    with console.status("[bold green]Invoking architect...") as status:
        prepare_architect()


    from rich.progress import Progress
    with Progress(transient=True) as progress:
        task = progress.add_task("[bold green]Launching tasks...", total=3)
        sleep(1)

        launch_tasks(progress.console)
        # simulate completion:
        for job in range(3):
            sleep(1.5)
            if job == 1:
                progress.console.log(log(f"[yellow]:hourglass: assignment/update", f"Assignment {job+1} timed-out."))
            else:
                progress.console.log(log(f"[green]:inbox_tray: assignment/update", f"Assignment {job+1} received."))
            progress.advance(task)

        sleep(1)



    console.log("Finished launching 3 tasks.")

    # with console.status("[bold green]Launching tasks...", spinner="earth") as status:
    #     launch_tasks()


start()
```

### PR DESCRIPTION
**Pull request by Etesam913**
## Summary
Many of the scripts are beautified using rich, why not add this to the logs.

Added a `RichHandler` object to the logging config to allow for clearer logging messages.

## Updated Video:
https://github.com/facebookresearch/Mephisto/pull/869#issuecomment-1198627769

## Video

https://user-images.githubusercontent.com/55665282/181590554-238b7ace-26c5-4fc1-9141-61873c3180aa.mov

## Duplicated Log Messages Issue (RESOLVED) https://github.com/facebookresearch/Mephisto/pull/869#issuecomment-1198627769
One thing that I want to fix is the duplicated log messages. You can see that each colorful log message has a uncolored version. 

I believe this is because the main logger is also logging the messages, which leads to duplicated logging. One way to fix this is to set 
```python
logger.propagate = False
```
This does work, however, it means that the messages won't get saved in the `outputs/<date>/scripts.log` file. This is a pretty big issue, so I didn't apply it. Removing the duplicated messages does make the logs much easier to read though.

## Resolves #386

### CODE REVIEW

**Review by JackUrb — COMMENTED** [review]
The duplicate logging is a symptom of [hydra's default logging](https://hydra.cc/docs/tutorials/basic/running_your_app/logging/). This _should_ be [configurable](https://hydra.cc/docs/configure_hydra/logging/).

With a bit of digging I'd suggest
1. Editing `hydra_config.py` to include a `LOGGING_OVERRIDE = {'override hydra/job_logging': 'mephisto_default'}` override constant.
2. Creating a file called `mephisto_default.yaml` in the specific folder `hydra_configs/hydra/job_logging`.
3. Include the `LOGGING_OVERRIDE` as the last value in a defaults list provided both to the standard `TaskConfig` and the version produced by `build_default_task_config`.

Now as far as what to do when you've gotten this far, I'm not 100% sure. Either the choice is going to be to make the `hydra_log_file.yaml` use the rich logger instead of the simple one (more likely), or to make it so that the `hydra_log_file` resembles the [disabled](https://github.com/facebookresearch/hydra/blob/1.2_branch/hydra/conf/hydra/hydra_logging/disabled.yaml) setup to block hydra's logs to console if that doesn't work.

**Comment by Etesam913** [comment]
Got it working 👍 I had to make sure that the hydra job_logging did not write results to console, but did write to file.

## New Video

https://user-images.githubusercontent.com/55665282/181636653-4c9528a9-0f90-415e-b463-b4027c9a4f06.mov

**Comment by Etesam913** [comment]
This is in mock_unit.py:
mock_unit py
Do we have to print this out given that it is already logged?

It seems a little extra.

**Comment by JackUrb** [comment]
> This is in `mock_unit.py`: mock_unit py Do we have to print this out given that it is already logged?
> 
> It seems a little extra.

My concern here is if someone has logging entirely disabled, and then doesn't get to see this message

**Comment by Etesam913** [comment]
Makes sense

File: mephisto/operations/hydra_config.py [hunk 1]
```diff
@@ -46,6 +46,7 @@ class TaskConfig:
     mephisto: MephistoConfig = MephistoConfig()
     task_dir: str = get_run_file_dir()
     num_tasks: int = 5
+    defaults = ["_self_", LOGGING_OVERRIDE]
```
Comment (JackUrb) [inline mephisto/operations/hydra_config.py:49]: You'll need to use:
```python
defaults: List[Any] = field(default_factory=lambda: ["_self_", LOGGING_OVERRIDE])
``` 
here or Hydra will throw for base `TaskConfig`s.

**Review by JackUrb — APPROVED** [review]
Thanks for this change! One small Hydra gotcha but otherwise I'm really excited to have improved, readable logs! Thanks for taking this on :)

File: test/core/test_live_runs.py [hunk 2]
```diff
@@ -124,6 +124,16 @@ def get_mock_run(self, blueprint, task_runner) -> LiveTaskRun:
...
 
+    def assert_server_subbed_in_time(self, server, timeout: int = 5) -> None:
+        start_time = time.time()
+        while (len(server.subs) == 0) and time.time() - start_time < timeout:
+            time.sleep(0.3)
+        self.assertEqual(
+            len(self.architect.server.subs),
+            1,
+            "MockServer doesn't see registered channel",
+        )
```
Comment (JackUrb) [inline test/core/test_live_runs.py:135]: ❤️

**Review by JackUrb — APPROVED** [review]

**Comment by pringshia** [comment]
Side thought - The duplication of logs here could be an opportunity for clean-up?

CleanShot 2022-08-01 at 11 37 41@2x

**Comment by JackUrb** [comment]
> Side thought - The duplication of logs here could be an opportunity for clean-up?
> 
> CleanShot 2022-08-01 at 11 37 41@2x

Etesam asked about this [above](https://github.com/facebookresearch/Mephisto/pull/869#issuecomment-1199378831). The difficulty here I suppose is that I wanted the link to _always_ appear, even if logging is disabled. A solution would be one that forces the print only if logging isn't active, but I don't know if there's any really elegant way to do that.

**Pull request merged**